### PR TITLE
PF Session 3 Assignments

### PR DESCRIPTION
**Pull request by meta-udit-saxena**

### CODE REVIEW

File: Maths.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	 * to compute the remainder of x divided by y.
+	 * 
+	 * @param x
+	 *            - first number where x>=0
+	 * @param y
+	 *            - second number where y>0
+	 * @return - remainder of (x,y)
+	 */
+	int rem(int x, int y) {
+		if (y <= 0) {
```
Comment (chirag-metacube) [inline Maths.java:21]: Value of y is already provided >0, No need to check

File: Maths.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	 *            first number where x>0
+	 * @param y
+	 *            second number where y>0
+	 * @return greatest common divisor of x,y
+	 */
+	int gcd(int x, int y) {
+		if (y == 0) {
+			return x;
+		}
+		int result = gcd(y, x % y);
```
Comment (chirag-metacube) [inline Maths.java:43]: can be return gcd(y, x % y);

File: Maths.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	/**
+	 * to find out the largest digit in the number x
+	 * 
+	 * @param x
+	 *            - given number where x>=0
+	 * @return max - largest digit present in that number
+	 */
+	int largestDigit(int x) {
+		// if given number is negative
+		x = Math.abs(x);
```
Comment (chirag-metacube) [inline Maths.java:57]: Value of x is already >=0

File: MathsTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import static org.junit.Assert.assertEquals;
```
Comment (chirag-metacube) [inline MathsTest.java:1]: Missing tests mentioned in assignment

rem(2, 1)
rem(2, 0)
rem(100, 3)

gcd(2, 1)
gcd(12, 18)
gcd(100, 3)

largestdigit(2)
largestdigit(1257369)
largestdigit(444)

File: QuickSortAlgorithm.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param array
+	 *            - integer array (Unsorted)
+	 * @param startIndex
+	 *            - starting index of array
+	 * @param endIndex
+	 *            - end index of array
+	 */
+	void quickSort(int[] array, int startIndex, int endIndex) {
+		try {
+			checkEmptyArray(array);
```
Comment (chirag-metacube) [inline QuickSortAlgorithm.java:53]: Not required, as start index and end index would be equal for empty array and below sorting logic will be skipped

File: QuickSortAlgorithm.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+			int pivot = array[endIndex];
+			for (int i = startIndex; i < endIndex; i++) {
+				if (array[i] < pivot) {
+					partitionIndex++;
+					swap(array, i, partitionIndex);
+				}
+			}
+			swap(array, endIndex, partitionIndex + 1);
+			return partitionIndex + 2;
+		} catch (Exception e) {
```
Comment (chirag-metacube) [inline QuickSortAlgorithm.java:35]: No need of exception handling, as already handled by parent function void quickSort(int[] array, int startIndex, int endIndex)

File: QuickSortAlgorithm.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param array
+	 *            - integer array (Unsorted)
+	 * @param startIndex
+	 *            - starting index of array
+	 * @param endIndex
+	 *            - end index of array
+	 * @return partitionIndex 
+	 * 			  -index from where array is divided into two part
+	 */
+	int returnPartitionIndex(int[] array, int startIndex, int endIndex) {
```
Comment (chirag-metacube) [inline QuickSortAlgorithm.java:22]: Method name can be getPartitionIndex instead of returnPartitionIndex

File: QuickSortAlgorithm.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param secondElementIndex
+	 *            - index of secondElement
+	 */
+	void swap(int[] array, int firstElementIndex, int secondElementIndex) {
+		try {
+			checkEmptyArray(array);
+			int temp = array[firstElementIndex];
+			array[firstElementIndex] = array[secondElementIndex];
+			array[secondElementIndex] = temp;
+		} catch (Exception e) {
```
Comment (chirag-metacube) [inline QuickSortAlgorithm.java:81]: No need to handle exception

File: QuickSortAlgorithm.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	/**
+	 * 
+	 * @param array
+	 *            - integer array
+	 * @throws Exception
+	 *             "Array is Empty"
+	 */
+	void checkEmptyArray(int[] array) throws Exception {
```
Comment (chirag-metacube) [inline QuickSortAlgorithm.java:93]: Instead of throwing exception this method should return a Boolean value, indicating whether array is empty

File: QuickSortAlgorithmTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import static org.junit.Assert.assertArrayEquals;
```
Comment (chirag-metacube) [inline QuickSortAlgorithmTest.java]: 1 test case missing from assignment.

File: QuickSortAlgorithmTest.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+		int actual = obj.returnPartitionIndex(array, 0, array.length - 1);
+		assertEquals(expected, actual);
+	}
+
+	/*
+	 * Boundary cases when array is NULL
+	 */
+
+	@Test
+	public void quickSortBoundaryTestCaseWhenArrayIsNull() {
```
Comment (chirag-metacube) [inline QuickSortAlgorithmTest.java:46]: No assert in this test case

File: QuickSortAlgorithmTest.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+		obj.quickSort(null, 0, 10);
+	}
+
+	@Test
+	public void swapBoundaryTestCaseWhenArrayIsNull() {
```
Comment (chirag-metacube) [inline QuickSortAlgorithmTest.java:51]: No assert in this test case

File: SearchAlgoritm.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	 * @param x
+	 *            - number to be found
+	 * @param startIndex
+	 *            - starting index of array
+	 * @return index of number or -1 if number not present
+	 */
+	int linearSearch(int[] array, int x, int startIndex) {
+		try {
+			// calling checkEmptyArray method from class QuickSortAlgorithm
+			new QuickSortAlgorithm().checkEmptyArray(array);
```
Comment (chirag-metacube) [inline SearchAlgoritm.java:24]: SearchAlgoritm should not call method from QuickSortAlgorithm, This method can be moved at a common place so that both classes can use it.

Comment (chirag-metacube) [reply SearchAlgoritm.java:24]: and no need to check checkEmptyArray here.

File: SearchAlgoritm.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	 *            - starting index of array
+	 * @param endIndex
+	 *            - end index of array
+	 * @return index of number or -1 if number not present
+	 */
+	int binarySearch(int[] array, int startIndex, int endIndex, int x) {
+		try {
+			// calling checkEmptyArray method from class QuickSortAlgorithm
+			new QuickSortAlgorithm().checkEmptyArray(array);
+			if (startIndex < 0 || endIndex < 0 || startIndex >= array.length
```
Comment (chirag-metacube) [inline SearchAlgoritm.java]: If condition can be replaced by endIndex>=startIndex

File: SearchAlgoritmTest.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
+import static org.junit.Assert.assertEquals;
```
Comment (chirag-metacube) [inline SearchAlgoritmTest.java:1]: 1 test case missing from assignments

File: SearchAlgoritmTest.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+		int actual = obj.linearSearch(array,7,0);
+		assertEquals(expected, actual);
+	}
+	
+	/*
+	 * Boundary cases when array is null
+	 */
+	@Test
+	public void linearSearchBoundaryTestCaseWhenArrayIsNull(){
+		int[] array =null;
```
Comment (chirag-metacube) [inline SearchAlgoritmTest.java:54]: Assert is missing from test case, This test case will not test anything.

**Review by chirag-metacube — APPROVED** [review]
